core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -657,6 +688,44 @@ class BlockManagerMasterEndpoint(
     }
   }
 
+  private def getShufflePushMergerLocations(
+      numMergersNeeded: Int,
+      hostsToFilter: Set[String]): Seq[BlockManagerId] = {
+    val blockManagersWithExecutors = blockManagerIdByExecutor.groupBy(_._2.host)
+      .mapValues(_.head).values.map(_._2).toSet
+    val filteredBlockManagersWithExecutors = blockManagersWithExecutors
+      .filterNot(x => hostsToFilter.contains(x.host))
+    val filteredMergersWithExecutors = filteredBlockManagersWithExecutors.map(
+      x => BlockManagerId(BlockManagerId.SHUFFLE_MERGER_IDENTIFIER,
+        x.host, externalShuffleServicePort))
+
+    // Enough mergers are available as part of active executors list
+    if (filteredMergersWithExecutors.size >= numMergersNeeded) {
+      filteredMergersWithExecutors.toSeq
+    } else {
+      // Delta mergers added from inactive mergers list to the active mergers list
+      val filteredMergersWithExecutorsHosts = filteredMergersWithExecutors.map(_.host)
+      val filteredMergersWithoutExecutors = shuffleMergerLocations.values
+        .filterNot(x => hostsToFilter.contains(x.host))
+        .filterNot(x => filteredMergersWithExecutorsHosts.contains(x.host))
+      val randomFilteredMergersLocations =
+        if (filteredMergersWithoutExecutors.size >
+          numMergersNeeded - filteredMergersWithExecutors.size) {
+          Utils.randomize(filteredMergersWithoutExecutors)
+        } else {
+          filteredMergersWithoutExecutors
+        }
+      filteredMergersWithExecutors.toSeq ++ randomFilteredMergersLocations
+        .take(numMergersNeeded - filteredMergersWithExecutors.size)
```

Comment:
```suggestion
      val randomFilteredMergersLocations =
        if (filteredMergersWithoutExecutors.size >
          numMergersNeeded - filteredMergersWithExecutors.size) {
          Utils.randomize(filteredMergersWithoutExecutors).take(numMergersNeeded - filteredMergersWithExecutors.size)
        } else {
          filteredMergersWithoutExecutors
        }
      filteredMergersWithExecutors.toSeq ++ randomFilteredMergersLocations
```

I mean we can do like this @Victsm @venkata91 . We can avoid unnecessary `take()` when `filteredMergersWithoutExecutors.size <= numMergersNeeded - filteredMergersWithExecutors.size`.